# Analyses adapted along the lines of:
# * https://github.com/pachterlab/aggregationDE/blob/master/SRP100701/R/tx_pipeline.R
# * https://github.com/pachterlab/aggregationDE/blob/master/SRP100701/R/GO_analysis.R
#
# As published in:
#   Yi, Lynn, Harold Pimentel, Nicolas L. Bray, and Lior Pachter. 2018.
#   “Gene-Level Differential Analysis at Transcript-Level Resolution.”
#   Genome Biology 19 (1): 53. https://doi.org/10.1186/s13059-018-1419-z.
#

library(topGO) # bioconductor-topgo
# gene-centric annotation of the Mus musculus genome
library(org.Mm.eg.db) # bioconductor-org.Mm.eg.db
# transcript-centric ensembl-based annotation of the Mus musculus genome
#library(TxDb.Mmusculus.UCSC.mm10.ensGene) # bioconductor-TxDb.Mmusculus.UCSC.mm10.ensGene
#library(GO.db) # GO Term explanations
library(stats)
library(tidyverse)
library(Rgraphviz) # bioconductor-rgraphviz
library(aggregation)


print('loading data')
tx_results <- readRDS("sleuth/diffexp/modelXYZ.diffexp.rds") %>%
  filter( !is.na(ens_gene))

# As an alternative to the Lancaster p-value aggregation, you can just load
# the gene-aggregated data as generated by sleuth
#g_results <- readRDS("sleuth/diffexp/modelXYZ.genes-aggregated.diffexp.rds")

# Lancaster p-value aggregation over genes
print('Performing Lancaster aggregation over genes')
tx_summarised <- tx_results %>%
  group_by(ens_gene) %>%
  summarise( lan = lancaster(pval, mean_obs),
             weight = sum(mean_obs, na.rm=TRUE)
            ) %>%
  mutate( lan_adjust = p.adjust(lan, method = 'BH')
          )

# list of genes of interest as a named vector (required for
# the topGO allGenes parameter), with two possibilities:
# 1. numerical vectors (a geneSel function would need to be
#    created and passed during topGOdata instatiation, so I
#    opt for option 2:)
# 2. factor vector with two levels (1 for "significant" gene,
#    0 for non-"significant" gene)
all_genes <- tx_summarised %>%
              mutate(gene_interesting = factor( if_else(lan_adjust < 0.05, 1, 0) ) ) %>%
              select(rowname = ens_gene, gene_interesting) %>%
              deframe()

# doing a classic Fisher test GO term enrichment analysis with topGO

tgd_classic <- new( "topGOdata",
                    ontology='BP',
                    allGenes = all_genes,
                    nodeSize=5,
                    annot=annFUN.org,
                    mapping="org.Mm.eg.db",
                    ID = "ensembl"
                    )
topGO_classic_Fisher <- runTest(tgd_classic,
                                algorithm = "classic",
                                statistic = "Fisher"
                                )
topGO_gen_table <- GenTable(tgd_classic,
                            Fisher_classic = topGO_classic_Fisher,
                            orderBy = Fisher_classic,
                            topNodes = length(topGO_classic_Fisher@score)
                            ) %>%
                    mutate(Fisher_classic = as.numeric(Fisher_classic))

showSigOfNodes(tgd_classic, score(topGO_classic_Fisher), firstSigNodes = 15, useInfo = 'all' )
